File: python/01_basic_chat_SAMBA.py
#!/usr/bin/env python3
"""
Example 1: Basic Chat with SambaNova

WHAT THIS DEMONSTRATES:
    - How to use SambaNova's enterprise AI platform
    - OpenAI-compatible API format with SambaNova
    - Using open-source Llama models via SambaNova
    - Enterprise-grade inference infrastructure
    - Standard library HTTP requests

WHAT YOU'LL LEARN:
    - SambaNova API endpoints
    - Using open models in production
    - OpenAI-compatible request format
    - Enterprise AI deployment patterns
    - Token usage with Llama models

PREREQUISITES:
    - Python 3.6 or higher
    - SAMBANOVA_API_KEY environment variable set
    - Internet connection

EXPECTED OUTPUT:
    - Full JSON response from SambaNova
    - Llama model's answer
    - Token usage statistics

Run with: python3 01_basic_chat_SAMBA.py
"""

import http.client
import json
import os

# ==============================================================================
# Step 1: Load SambaNova API credentials
# ==============================================================================

api_key = os.environ.get('SAMBANOVA_API_KEY')

if not api_key:
    print("Error: SAMBANOVA_API_KEY environment variable not set")
    print("Run: export SAMBANOVA_API_KEY='...'")
    print("Get your key from: https://sambanova.ai")
    exit(1)

# ==============================================================================
# Step 2: Build the request payload (OpenAI-compatible)
# ==============================================================================

payload = {
    "model": "Meta-Llama-3.1-8B-Instruct",  # Open-source Llama model
    "messages": [
        {
            "role": "user",
            "content": "What is the capital of Switzerland?"
        }
    ],
    "temperature": 0.7,  # Creativity level
    "max_tokens": 100    # Response length limit
}

payload_json = json.dumps(payload)

# ==============================================================================
# Step 3: Connect to SambaNova API
# ==============================================================================

# SambaNova provides OpenAI-compatible endpoints
conn = http.client.HTTPSConnection("api.sambanova.ai")

# ==============================================================================
# Step 4: Send request with SambaNova headers
# ==============================================================================

# SambaNova uses OpenAI-style Bearer authentication
headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {api_key}'
}

try:
    # Use OpenAI-compatible endpoint
    conn.request("POST", "/v1/chat/completions", payload_json, headers)

    # Get the response
    response = conn.getresponse()
    response_data = response.read().decode('utf-8')

    # Parse JSON response
    response_json = json.loads(response_data)

    # ==============================================================================
    # Step 5: Display the results
    # ==============================================================================

    print("Full SambaNova Response:")
    print(json.dumps(response_json, indent=2))

    # Extract and display the answer (OpenAI-compatible format)
    if 'choices' in response_json and len(response_json['choices']) > 0:
        answer = response_json['choices'][0]['message']['content']
        print("\n" + "="*50)
        print("Llama's Answer (via SambaNova):")
        print(answer)

    # Show token usage
    if 'usage' in response_json:
        usage = response_json['usage']
        print("\n" + "="*50)
        print("Token Usage:")
        print(f"  Prompt tokens: {usage.get('prompt_tokens', 0)}")
        print(f"  Completion tokens: {usage.get('completion_tokens', 0)}")
        print(f"  Total tokens: {usage.get('total_tokens', 0)}")

    # Display model info
    if 'model' in response_json:
        print(f"\nModel used: {response_json['model']}")
        print("(Open-source Llama model via SambaNova)")

    # Check for errors
    if 'error' in response_json:
        print("\nAPI Error:")
        print(json.dumps(response_json['error'], indent=2))

except Exception as e:
    print(f"Request failed: {e}")
    print("\nTroubleshooting:")
    print("1. Check your internet connection")
    print("2. Verify your SambaNova API key")
    print("3. Check SambaNova service status")

finally:
    conn.close()

# ==============================================================================
# SambaNova API Notes:
# ==============================================================================
#
# Key features:
#   - OpenAI-compatible format (easy migration)
#   - Focus on open-source models (Llama, Mistral)
#   - Enterprise-grade infrastructure
#   - High-performance custom hardware
#   - Privacy-focused deployments
#
# Available models:
#   - Meta-Llama-3.1-8B-Instruct: Fast, efficient
#   - Meta-Llama-3.1-70B-Instruct: More capable
#   - Meta-Llama-3.1-405B-Instruct: State-of-the-art
#
# Use cases:
#   - Enterprise deployments requiring open models
#   - Privacy-sensitive applications
#   - High-throughput production workloads
#   - Cost-effective scaling
#
# Advantages:
#   - No vendor lock-in (open models)
#   - Predictable performance
#   - Enterprise support available
#

File: nodejs/01_basic_chat_SAMBA.js
/**
 * ============================================================================
 * Example 1: Basic Chat with SambaNova - Enterprise Open Models
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 * - How to make a SambaNova API call using Node.js
 * - SambaNova's OpenAI-compatible format
 * - Using open-source models in enterprise settings
 * - No external dependencies needed!
 *
 * WHAT YOU'LL LEARN:
 * - Building HTTPS requests for SambaNova
 * - Working with Llama models via SambaNova
 * - OpenAI-compatible API format
 * - Enterprise-grade open model deployment
 *
 * PREREQUISITES:
 * - Node.js installed (v14 or higher)
 * - SAMBANOVA_API_KEY environment variable set
 *   Run: export SAMBANOVA_API_KEY="..."
 *
 * EXPECTED OUTPUT:
 * - Full JSON response from SambaNova
 * - The AI's answer using Llama model
 * - Token usage statistics
 *
 * RUN THIS:
 * node 01_basic_chat_SAMBA.js
 * ============================================================================
 */

// Step 1: Import the built-in HTTPS module
const https = require('https');

// Step 2: Prepare the SambaNova request payload
// SambaNova uses OpenAI-compatible format
const data = JSON.stringify({
  model: "Meta-Llama-3.1-8B-Instruct",         // Open-source Llama model
  messages: [
    {
      role: "user",
      content: "What is the capital of Switzerland?"
    }
  ],
  temperature: 0.7,   // Creativity level
  max_tokens: 100     // Maximum response length
});

// Step 3: Configure the SambaNova HTTP request
// SambaNova uses OpenAI-compatible endpoints
const options = {
  hostname: 'api.sambanova.ai',                // SambaNova's API server
  port: 443,
  path: '/v1/chat/completions',                // OpenAI-compatible endpoint
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${process.env.SAMBANOVA_API_KEY}`,  // Standard Bearer auth
    'Content-Length': Buffer.byteLength(data)
  }
};

// Step 4: Create and send the request
const req = https.request(options, (res) => {
  let responseData = '';

  // Accumulate response chunks
  res.on('data', (chunk) => {
    responseData += chunk;
  });

  // Process complete response
  res.on('end', () => {
    try {
      const parsed = JSON.parse(responseData);

      // Display full response
      console.log('Full SambaNova Response:');
      console.log(JSON.stringify(parsed, null, 2));

      // Extract the AI's answer (OpenAI-compatible format)
      if (parsed.choices && parsed.choices[0]) {
        console.log('\nLlama Answer:');
        console.log(parsed.choices[0].message.content);
      }

      // Show token usage
      if (parsed.usage) {
        console.log('\nToken Usage:');
        console.log(`Prompt: ${parsed.usage.prompt_tokens}`);
        console.log(`Completion: ${parsed.usage.completion_tokens}`);
        console.log(`Total: ${parsed.usage.total_tokens}`);
      }

      // Check for errors
      if (parsed.error) {
        console.error('API Error:', parsed.error.message);
      }

    } catch (error) {
      console.error('Error parsing response:', error.message);
      console.log('Raw response:', responseData);
    }
  });
});

// Handle request errors
req.on('error', (error) => {
  console.error('Request failed:', error.message);

  if (error.message.includes('ENOTFOUND')) {
    console.log('Check your internet connection');
  }
  if (!process.env.SAMBANOVA_API_KEY) {
    console.log('Make sure SAMBANOVA_API_KEY is set');
  }
});

// Send the request
req.write(data);
req.end();

// SambaNova API Notes:
// - OpenAI-compatible format (easy migration)
// - Focuses on open-source models (Llama, Mistral, etc.)
// - Enterprise-grade infrastructure
// - High-performance inference optimized hardware
// - Models: Llama 3.1 variants (7B, 70B, 405B)
// - Good for privacy-conscious deployments